### ISSUE TITLE
Make child window logs silenceable

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/e2e-testing#46]
I'd like to be able to silence logging in the child window. In #32 we discussed a possible solution for errors by capturing them with window.onerror. For simple logs I think the only way to override the console object or wrap all of its methods with a silencer.

**Comment by inf3rno** [comment]
I think this won't be a problem. Karma did not capture any log from the child window. I don't know why I expected it would. It is very logical that Karma captures logs of the parent window only, because only that is connected to the karma server. The interesting thing here that you cannot reach the console of a different window. I tried to use the console of the parent window, but it just did not work.

**Comment by inf3rno** [comment]
parent
```js
    this.win = win;
    window.logger = {
        log: function (){
            return console.log.apply(console, arguments);
        }
    };
    this.win.logger = window.logger;
```

child
```js
addAllEventListeners("head", window, window.logger.log); // does not work
addAllEventListeners("head", window, window.opener.logger.log); // works
```

It works only if it is called through window.opener. Otherwise it doesn't. We cannot delegate any function to the child, which does logging. I am not sure whether we have the native console by karma or a wrapped one. If it is a wrapped one, we could publish the object instead of a function on the opener. It would not make a real difference though. I don't like to publish functions or objects to global namespace, but I guess I have to do it now.

**Comment by inf3rno** [comment]
I checked, the karma console is not wrapped. Doing `window.logger = console` did not work.

**Comment by inf3rno** [comment]
Okay I figured this out. If I add it from the parent while the document is loading, then it works. If I add it before the document was loaded, then it works only if I call it through window.opener. So I guess the variable is removed by every page load. Strange that it does not give some kind of error. And as far as I remember window variables stay there. Still somehow the connection is lost to the parent window console.

I added `this.win.console = window.console;` to the loading event and it works perfectly. :-)))

**Comment by inf3rno** [comment]
According to this: https://stackoverflow.com/questions/45813175/how-can-i-console-log-to-parent-window the upper code should not work. I need to check it in browser.

**Comment by inf3rno** [comment]
I checked, it is problematic and it is related to #34 . I tested unload + next tick and script running with this code:
```js
if (window.opener){
	window.console = window.opener.console;
	console.log("child script", window.x, new Date() - 0);
	window.onload = function (){
		console.log("child load", window.x, new Date() - 0);
	};
	setTimeout(function (){
		console.log("child 1", window.x, new Date() - 0);
	}, 1);
	setTimeout(function (){
		console.log("child 500", window.x, new Date() - 0);
	}, 500);
	setTimeout(function (){
		console.log("child 2000", window.x, new Date() - 0);
	}, 2000);
}
else {
	var uuid = "win." + Math.random();
	var win = window.open("about:blank", uuid, "menubar=yes,location=yes,resizable=yes,scrollbars=yes,status=yes");
	win.onunload = function (){
		console.log("previous unload", new Date() - 0);
		setTimeout(function (){
			console.log("previous unload + next tick", new Date() - 0);
		}, 0);
	};
	win.location.href = location.href;
}
```

**Comment by inf3rno** [comment]
The results are very bad:

```
Chrome:

	previous unload 1534430178714
	child script 1534430178718
	child load 1534430178718
	child 1 1534430178719
	child 500 1534430179218
	previous unload + next tick 1534430179230
	child 2000 1534430181230

Firefox:

	previous unload 1534430256232 b.html:22:3
	previous unload + next tick 1534430256233 b.html:24:4
	child script 1534430256235 b.html:4:2
	child load 1534430256236 b.html:6:3
	child 1 1534430256238 b.html:9:3
	child 500 1534430256736 b.html:12:3
	child 2000 1534430258236
	
Opera:

	previous unload 1534430291226
	previous unload + next tick 1534430291230
	child script 1534430291254
	child load 1534430291254
	child 1 1534430291264
	child 500 1534430291753
	child 2000 1534430294231
	
IE11:

	previous unload 1534430325362
	previous unload + next tick 1534430325369
	child script 1534430325372
	child load 1534430325380
	child 1 1534430325384
	child 500 1534430325885
	child 2000 1534430327381
```

**Comment by inf3rno** [comment]
For Karma + Chrome + security.disabled here are the results:

```
LOG: '0.118200', 'parent', 'unload', 'about:blank'
LOG: '0.123900', 'parent', 'unload+nexttick', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.011900', 'parent', 'loading', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.013400', 'child: head', 'loading', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.013700', 'child: body begin', 'loading', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.013900', 'child: body end', 'loading', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.014100', 'parent', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'interactive'
LOG: '0.014100', 'child: head', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'interactive'
LOG: '0.014200', 'child: body begin', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'interactive'
LOG: '0.014200', 'child: body end', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'interactive'
LOG: '0.014400', 'parent', 'document.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.014400', 'child: head', 'document.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.014500', 'child: body begin', 'document.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.014500', 'child: body end', 'document.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.014600', 'parent', 'win.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.014600', 'child: head', 'win.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.014600', 'child: body begin', 'win.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.014800', 'child: body end', 'win.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.014900', 'parent', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'complete'
LOG: '0.014900', 'child: head', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'complete'
LOG: '0.014900', 'child: body begin', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'complete'
LOG: '0.015100', 'child: body end', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'complete'
LOG: '0.015100', 'child: body end', 'win.onload', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.015200', 'parent', 'win.addEventListener:load', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.015300', 'child: head', 'win.addEventListener:load', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.015300', 'child: body begin', 'win.addEventListener:load', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '0.015400', 'child: body end', 'win.addEventListener:load', 'http://localhost:9876/base/test/karma.static/helloworld.html'
```

**Comment by inf3rno** [comment]
It does not depend on loading the favicon.

**Comment by inf3rno** [comment]
I'll create a similar Karma log and report this as a bug by both Opera and Chrome.

**Comment by inf3rno** [comment]
I ran the original logger script on Karma and several browsers too

```
Chrome
	previous unload 1534434676622
	b.html:5 child script 1534434676625
	b.html:7 child load 1534434676626
	b.html:10 child 1 1534434676627
	b.html:13 child 500 1534434677127
	b.html:33 polling document change 1534434677259
	b.html:25 previous unload + next tick 1534434677259
	b.html:16 child 2000 1534434679258
	
Firefox
	previous unload 1534434704892 b.html:23:3
	previous unload + next tick 1534434704894 b.html:25:4
	polling document change 1534434704894 b.html:33:3
	child script 1534434704895 b.html:5:2
	child load 1534434704897 b.html:7:3
	child 1 1534434704899 b.html:10:3
	child 500 1534434705396 b.html:13:3
	child 2000 1534434706896
	
Opera
	previous unload 1534434730667
	favicon.ico:1 Failed to load resource: the server responded with a status of 404 (Not Found)
	b.html:5 child script 1534434730678
	b.html:7 child load 1534434730679
	b.html:10 child 1 1534434730684
	b.html:13 child 500 1534434731179
	b.html:33 polling document change 1534434731260
	b.html:25 previous unload + next tick 1534434731260
	b.html:16 child 2000 1534434732679
	
IE11
	previous unload 1534434787694
	polling document change 1534434787700
	previous unload + next tick 1534434787701
	child script 1534434787702
	child load 1534434787706
	child 1 1534434787712
	child 500 1534434788207
	child 2000 1534434789708

Karma + Chrome
	LOG: 'previous unload', 1534448614483
	LOG: 'previous unload + next tick', 1534448614487
	LOG: 'polling document change', 1534448614487
	LOG: 'child script', 1534448614489
	LOG: 'child load', 1534448614489
	LOG: 'child 1', 1534448614489
	LOG: 'child 500', 1534448614989
	LOG: 'child 2000', 1534448616489

Karma + Firefox
	LOG: 'previous unload', 1534448725685
	LOG: 'polling document change', 1534448725694
	LOG: 'previous unload + next tick', 1534448725695
	LOG: 'child script', 1534448725708
	LOG: 'child load', 1534448725715
	LOG: 'child 1', 1534448725720
	LOG: 'child 500', 1534448726208
	LOG: 'child 2000', 1534448727712
	
Karma + IE10
	LOG: 'previous unload', 1534449064850
	LOG: 'child script', 1534449064857
	LOG: 'child load', 1534449064862
	LOG: 'polling document change', 1534449064865
	LOG: 'previous unload + next tick', 1534449064866
	LOG: 'child 1', 1534449064868
	LOG: 'child 500', 1534449065367
	LOG: 'child 2000', 1534449066864

Karma + IE11
	LOG: 'previous unload', 1534449118700
	LOG: 'polling document change', 1534449118704
	LOG: 'previous unload + next tick', 1534449118706
	LOG: 'child script', 1534449118706
	LOG: 'child load', 1534449118710
	LOG: 'child 1', 1534449118714
	LOG: 'child 500', 1534449119214
	LOG: 'child 2000', 1534449120712

Karma + Opera
	LOG: 'previous unload', 1534449158360
	LOG: 'child script', 1534449158371
	LOG: 'child load', 1534449158372
	LOG: 'child 1', 1534449158373
	LOG: 'polling document change', 1534449158608
	LOG: 'previous unload + next tick', 1534449158608
	LOG: 'child 500', 1534449158872
	LOG: 'child 2000', 1534449160372
	
Karma + PhantomJS
	LOG: 'previous unload', 1534449204814
	LOG: 'child script', 1534449204815
	LOG: 'child load', 1534449204816
	LOG: 'polling document change', 1534449204816
	LOG: 'previous unload + next tick', 1534449204819
	LOG: 'child 1', 1534449204820
	LOG: 'child 500', 1534449205326
	LOG: 'child 2000', 1534449206824
```

**Comment by inf3rno** [comment]
I ran a few other tests and I am certain now that capturing the document in loading state is not possible in IE10-11, Opera, PhantomJS and Chrome. On the other hand it is possible with Firefox and Chrome started with Karma runner. So if I want to have something general, then all I can do is an after:load event, which fires after the page is loaded, sometimes even 750 msec later. If I want to support only Chrome + Karma, then I can capture the page in a loading state and add an event listener before the script in the head runs. But it is not something safe, because Chrome on manual does not behave the same, so maybe after a version upgrade it will stop working. Firefox would be ok, but it does not have a CLI configurable security disabled mode, so all I could do is adding an injection server to the tested domain, which would do the navigation and which would be able to communicate with the tests hosted on the Karma server with postMessage.

The conclusion in this topic that they are silenced by default, and we need to add a `window.console = window.opener.console` to the new document in a loading state, which is possible only with Chrome + Karma and with Firefox. We can partially implement this by changing the console only after load, but that would miss a lot of logs.

**Comment by inf3rno** [comment]
We left out mutation observers etc... As far as I can tell it is some kind of polling on the document object, so it won't be that useful in this situation. But there are many object change observers, which we can check. Even legacy ones might work, who knows...

**Comment by inf3rno** [comment]
We left out mutation observers etc... As far as I can tell it is some kind of polling on the document object, so it won't be that useful in this situation. But there are many object change observers, which we can check. Even legacy ones might work, who knows... Ofc there is 5% chance I can workaround this, but maybe I have luck. :-)

**Comment by inf3rno** [comment]
We left out mutation observers etc... As far as I can tell it is some kind of polling on the document object, so it won't be that useful in this situation. But there are many object change observers, which we can check. Even legacy ones might work, who knows... Ofc there is 5% chance I can workaround this, but maybe I have luck. :-)

**Comment by inf3rno** [comment]
I think I can close this, because they are silenced by default. I need a way to listen to them though, so I think I'll open a new issue for that. The struggling with navigation events is another issue too.
